ZConfig/tests/test_info.py

```diff
@@ -0,0 +1,228 @@
+
+    def test_subtypes(self):
+
+        t = AbstractType('name')
+        self.assertFalse(t.hassubtype('foo'))
+        self.assertEqual([], list(t.getsubtypenames()))
+
+        self.name = 'foo'
+        t.addsubtype(self)
+        self.assertTrue(t.hassubtype('foo'))
+
+class SectionTypeTestCase(unittest.TestCase):
+
+    def make_one(self, name='', keytype=None, valuetype=None,
+                 datatype=None, registry={}, types=None):
+        return SectionType(name, keytype, valuetype, datatype, registry, types)
+
+    def test_getinfo_no_key(self):
+        info = self.make_one()
+        self.assertRaisesRegexp(ConfigurationError,
+                                "cannot match a key without a name",
+                                info.getinfo,
+                                None)
+
+    def test_required_types_with_name(self):
+        info = self.make_one(name='foo')
+        self.assertEqual(['foo'], info.getrequiredtypes())
+
+    def test_getsectioninfo(self):
+        class MockChild(object):
+            _issection = False
+            def issection(self):
+                return self._issection
+        child = MockChild()
+
+        info = self.make_one()
+
+        info._children.append(('foo', child))
+
+        self.assertRaisesRegexp(ConfigurationError,
+                                'already in use for key',
+                                info.getsectioninfo,
+                                None, 'foo')
+
+        self.assertRaisesRegexp(ConfigurationError,
+                                'no matching section',
+                                info.getsectioninfo,
+                                None, 'baz')
+
+class SchemaTypeTestCase(unittest.TestCase):
+
+    def test_various(self):
+        class Mock(object):
+            pass
+
+        mock = Mock()
+        schema = SchemaType(None, None, None, None, 'url', {})
+
+        mock.name = 'name'
+        schema.addtype(mock)
+        with self.assertRaises(SchemaError):
+            schema.addtype(mock)
+
+        self.assertTrue(schema.allowUnnamed())
+        self.assertFalse(schema.isAllowedName(None))
+
+        with self.assertRaises(SchemaError):
+            schema.deriveSectionType(schema, None, None, None, None)
+
+        schema.addComponent('name')
+        self.assertRaisesRegexp(SchemaError,
+                                'already have component',
+                                schema.addComponent,
+                                'name')
+
+def test_suite():
+    suite = unittest.makeSuite(UnboundTestCase)
+    suite.addTest(unittest.makeSuite(BaseInfoTestCase))
+    suite.addTest(unittest.makeSuite(BaseKeyInfoTestCase))
+    suite.addTest(unittest.makeSuite(KeyInfoTestCase))
+    suite.addTest(unittest.makeSuite(SectionInfoTestCase))
+    suite.addTest(unittest.makeSuite(AbstractTypeTestCase))
+    suite.addTest(unittest.makeSuite(SectionTypeTestCase))
+    suite.addTest(unittest.makeSuite(SchemaTypeTestCase))
+    return suite
```

Comment:
I don't recall where I learned this trick, but I recall getting burnt a few times in Zopeland when I (or somebody else) wrote a new test class and then forgot the corresponding `makeSuite()` bit.  Of course, measuring test coverage protects from that to some extent.

I also have doubts whether that refactoring belongs to this already-very-large PR (especially if we want to convert all the files in one go, rather opt for a more incremental approach and convert just those that are touched in this PR).

I have this habit where when I'm reviewing code I also tend to look for possible improvements in the surrounding codebase, and also for opportunities to share knowledge.  But then I forget to state that that's what I'm doing, rather than asking for changes in this particular PR.